src/main/host/syscall/futex.c

```diff
@@ -0,0 +1,164 @@
+    int futex_op = args->args[1].as_i64;
+    int val = args->args[2].as_i64;
+    PluginPtr timeoutptr = args->args[3].as_ptr; // const struct timespec*, or uint32_t
+    PluginPtr uaddr2ptr = args->args[4].as_ptr;  // int*
+    int val3 = args->args[5].as_i64;
+
+    const int possible_options = FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME;
+    int options = futex_op & possible_options;
+    int operation = futex_op & ~possible_options;
+
+    debug("futex called with addr=%p op=%i (operation=%i and options=%i) and val=%i",
+          (void*)uaddrptr.val, futex_op, operation, options, val);
+
+    // Futexes are 32 bits on all platforms
+    void* faddr = (uint32_t*)uaddrptr.val;
+    uint32_t expectedVal = (uint32_t)val;
+
+    // futex addr cannot be NULL
+    if (!faddr) {
+        debug("Futex addr cannot be NULL");
+        return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = -EFAULT};
+    }
+
+    int result = 0;
+
+    switch (operation) {
+        case FUTEX_WAIT: {
+            debug("Handling FUTEX_WAIT operation %i", operation);
+
+            // This is a new wait operation on the futex for this thread.
+            // Check if a timeout was given in the syscall args.
+            const struct timespec* timeout;
+            if (timeoutptr.val) {
+                timeout =
+                    process_getReadablePtr(sys->process, sys->thread, timeoutptr, sizeof(*timeout));
+                // Bounds checking
+                if (!(timeout->tv_nsec >= 0 && timeout->tv_nsec <= 999999999)) {
+                    debug("A futex timeout was given, but the nanos value is out of range");
+                    result = -EINVAL;
+                    break;
+                }
+            } else {
+                timeout = NULL;
+            }
+
+            // TODO: does this load/compare need to be done atomically?
+            // `man 2 futex`: blocking via a futex is an atomic compare-and-block operation
+            const uint32_t* futexVal =
+                process_getReadablePtr(sys->process, sys->thread, uaddrptr, sizeof(uint32_t));
+
+            debug("Futex value is %" PRIu32 ", expected value is %" PRIu32, *futexVal, expectedVal);
+            if (*futexVal != expectedVal) {
+                debug("Futex values don't match, try again later");
+                result = -EAGAIN;
+                break;
+            }
+
+            // Check if we already have a futex
+            FutexTable* ftable = host_getFutexTable(sys->host);
+            Futex* futex = futextable_get(ftable, faddr);
+
+            if (_syscallhandler_wasBlocked(sys)) {
+                utility_assert(futex != NULL);
+
+                // We already blocked on wait, so this is either a timeout or wakeup
+                if (timeout && _syscallhandler_didListenTimeoutExpire(sys)) {
+                    // Timeout while waiting for a wakeup
+                    debug("Futex %p timeout out while waiting", faddr);
```

Comment:
Is there a possible race condition here if thread A does a WAKE-1 operation, waking thread B, but when it gets here finds the timeout *just* expired? Or does Shadow's scheduling guarantee that can't happen?

Reply:
I think we can consider whichever callback gets executed first is the thing that happened first. In any case, the futex man page states that spurious wakeups can occur, so I wouldn't call it a race condition even if I'm wrong.